test/hasManyThroughUpsert.test.js

```diff
@@ -0,0 +1,179 @@
+var request = require("supertest");
+var loopback = require('loopback')
+var expect = require('chai').expect;
+var JSONAPIComponent = require('../')
+var RSVP = require('rsvp')
+
+var app
+var Movie, Category, MovieCategoryAssoc;
+
+describe('hasManyThrough upsert', function () {
+  beforeEach(function (done) {
+    app = loopback()
+    app.set('legacyExplorer', false)
+    var ds = loopback.createDataSource('memory')
+    // create models
+    Movie = ds.createModel('movie', {
+      id: {type: Number, id: true},
+      name: String
+    })
+
+    Category = ds.createModel('category', {
+      id: {type: Number, id: true},
+      name: String
+    })
+
+    MovieCategoryAssoc = ds.createModel('movieCategoryAssoc', {
+      id: {type: Number, id: true}
+    })
+
+    // add models
+    app.model(Movie)
+    app.model(Category)
+    app.model(MovieCategoryAssoc)
+
+    // set up relationships
+    Movie.hasMany(Category, { through: MovieCategoryAssoc })
+    Category.hasMany(Movie, { through: MovieCategoryAssoc })
+
+    MovieCategoryAssoc.belongsTo(Movie)
+    MovieCategoryAssoc.belongsTo(Category)
+    makeData().then(done())
```

Comment:
Just to note this should be es5 rather than es6 syntax but the point stands.

```js
makeData().then(function (done) {
  done()
}).catch(function (err) {
  done(err)
})
```